File: C++/1. Two Pointer Techinique/1_target_sum.cpp
// reference: https://leetcode.com/problems/two-sum/

#include <bits/stdc++.h>
using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    vector<pair<int, int>> numsWithIndices(nums.size());
    for (int i = 0; i < nums.size(); ++i) {
        numsWithIndices[i] = make_pair(nums[i], i);
    }

    sort(numsWithIndices.begin(), numsWithIndices.end());

    vector<int> result;
    int left = 0;
    int right = nums.size() - 1;

    while (left < right) {
        int currentSum = numsWithIndices[left].first + numsWithIndices[right].first;
        if (currentSum == target) {
            result.push_back(numsWithIndices[left].second);
            result.push_back(numsWithIndices[right].second);
            return result;
        } else if (currentSum < target) {
            left++;
        } else {
            right--;
        }
    }

    return result;
}

int main() {
    vector<int> nums = {3, 4, 1};
    int target = 7;
    vector<int> result = twoSum(nums, target);

    for (int num : result) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: JavaScript/1. Two Pointer Techinique/1_target_sum.js
// reference: https://leetcode.com/problems/two-sum/

function two_sum(nums, target) {
    const nums_with_indices = nums.map((num, idx) => ({num, idx}));
    nums_with_indices.sort((i, j) => i.num - j.num);

    let result = [];
    let left = 0;
    let right = nums.length - 1;

    while (left < right) {
        const current_sum = nums_with_indices[left].num + nums_with_indices[right].num;
        if (current_sum === target) {
            result.push(nums_with_indices[left].idx);
            result.push(nums_with_indices[right].idx);
            return result;
        }
        else if (current_sum < target)
            left++;
        else 
            right--;
    }    
    return [];
}

console.log(two_sum([3, 4, 1], 7));

// Try out the following test cases
// nums = [2, 7, 11, 15] target = 9
// nums = [3, 2, 4] target = 6
// nums = [3, 3] target = 6
// nums = [-1, -2, -3, -4, -5] target = -8
// nums = [0, 0, 0, 0, 1, 2] target = 2
// nums = [1, 2, 3, 4, 5] target = 9
// nums = [1, 2, 3, 4, 5] target = 1
// nums = [1, 2, 3, 4, 5] target = 10
// nums = [-5, -3, -1, 0, 2, 4, 6] target = 1
// nums = [-5, -3, -1, 0, 2, 4, 6] target = -10